src/charm.py

```diff
@@ -0,0 +1,293 @@
+from jinja2 import Environment, FileSystemLoader
+
+logger = logging.getLogger(__name__)
+
+SERVICE = "snap.aar.daemon.service"
+UA_STATUS_PATH = "/var/lib/ubuntu-advantage/status.json"
+SNAP_COMMON_PATH = Path("/var/snap/aar/common")
+AAR_CONFIG_PATH = SNAP_COMMON_PATH / "conf/main.yaml"
+SNAP_NAME = "aar"
+
+AAR_CERT_BASE_PATH = SNAP_COMMON_PATH / "certs"
+CLIENT_CERT_PATH = AAR_CERT_BASE_PATH / "clients"
+PUBLISHERS_CERT_PATH = AAR_CERT_BASE_PATH / "publishers"
+AAR_SERVER_CERT_PATH = AAR_CERT_BASE_PATH / "server.crt"
+AAR_SERVER_KEY_PATH = AAR_CERT_BASE_PATH / "server.key"
+
+
+class AARCharm(ops.CharmBase):
+    """Charmed Operator to deploy AAR - Anbox Application Registry."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.client_relation_joined, self._on_aar_joined)
+        self.framework.observe(self.on.publisher_relation_joined, self._on_aar_joined)
+
+        self.framework.observe(self.on.client_relation_changed, self._on_aar_changed)
+        self.framework.observe(self.on.publisher_relation_changed, self._on_aar_changed)
+
+    def _on_aar_joined(self, event: ops.RelationJoinedEvent):
+        with open(AAR_SERVER_CERT_PATH, "r") as f:
+            cert = f.read()
+        if not cert:
+            self.unit.status = ops.BlockedStatus("No registry certificate")
+            return
+
+        listen_address = self.private_ip
+        location = self.config.get("location")
+        if location:
+            listen_address = location
+
+        unit_data = event.relation.data[self.unit]
+        unit_data["certificate"] = cert
+        unit_data["fingerprint"] = self._get_fingerprint(cert)
+        unit_data["ip"] = listen_address
+        unit_data["port"] = str(self.config.get("port", ""))
+
+    def _get_fingerprint(self, cert: str) -> str:
+        x509_cert = x509.load_pem_x509_certificate(cert.encode())
+        return x509_cert.fingerprint(hashes.SHA256()).hex()
+
+    def _remove_all_certificates(self):
+        """Remove all old certificates held by AAR.
+
+        All client certificates are kept by juju in the relation until the hook
+        fails, so every time a new client is registered, the previous certificates
+        will be added again. To avoid duplicates and properly take into account
+        units that departed, we remove all certificates and add the ones that
+        are still active again.
+        """
+        shutil.rmtree(CLIENT_CERT_PATH)
+        shutil.rmtree(PUBLISHERS_CERT_PATH)
+        os.makedirs(CLIENT_CERT_PATH, 0o700, exist_ok=True)
+        os.makedirs(PUBLISHERS_CERT_PATH, 0o700, exist_ok=True)
+
+    def _on_aar_changed(self, event: ops.RelationChangedEvent):
+        self.unit.status = ops.MaintenanceStatus("Confguring new AMS Client")
```

Comment:
We mix lower/upper case for the beginning of the status message, if there is no recommendation we should at least be consistent.